lightning/src/events/mod.rs

```diff
@@ -524,6 +524,25 @@ impl_writeable_tlv_based_enum_upgradable!(HTLCHandlingType,
 	},
 );
 
+/// The reason for HTLC failures in [`Event::HTLCHandlingFailed`].
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum HTLCHandlingFailureReason {
+	/// The forwarded HTLC was failed back by the downstream node with an encrypted error reason.
+	Downstream,
+	/// The HTLC was failed locally by our node.
+	Local {
+		/// The reason that our node chose to fail the HTLC.
+		reason: LocalHTLCFailureReason,
```

Comment:
Thoughts on renaming `LocalHTLCFailureReason` to `HTLCFailureReason`? Then we can have `Downstream` just be a variant within the big enum and use the same big enum for the `handling_failure` field. It's a bit awkward that `HTLCFailReason` is another type in the codebase, but that could either be renamed to `RawOnionFailure` (no pun intended) or something, or punted for later.

Reply:
> Then we can have Downstream just be a variant within the big enum and use the same big enum for the handling_failure field. It

Doesn't this get a bit mesy when it comes to mapping to B04 codes because `Downstream` wouldn't have a value?

Reply:
Ah, I was missing that. Good point indeed.

Reply:
Echo your sentiment though, since`HTLCHandlingFailureReason`  is pretty much just an API-facing version of `HTLCFailReason` (hiding all the onion details the API doesn't care about) it is confusing to have two pretty similar structs. 

Might still be worth doing a rename of `HTLCFailReason`?

Reply:
Yeah, I tend to agree that a rename would still be nice. Looks like ~70 occurrences of `HTLCFailReason` currently, I think if you're up for it then go for it!